File: 0141-linked-list-cycle/0141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode pre = head;
        ListNode cur = head;
        boolean flag = false;
        while (cur != null && cur.next != null) {
            pre = pre.next;
            cur = cur.next.next;
            if (pre == cur) {
                flag = true;
                break;
            }
        }
        return flag;
    }
}

File: 0141-linked-list-cycle/0141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr) {
            return false;
        }
        ListNode* p1 = head;
        ListNode* p2 = head;
        while (p2 != nullptr && p2->next != nullptr) {
            p1 = p1->next;
            p2 = p2->next->next;
            if (p2 == nullptr || p2->next == nullptr) {
                return false;
            }
            if (p1 == p2) {
                return true;
            }
        }
        return false;
    }
};